File: src/main/java/com/pubnub/api/workers/SubscribeMessageProcessor.java
package com.pubnub.api.workers;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.pubnub.api.PNConfiguration;
import com.pubnub.api.PubNub;
import com.pubnub.api.PubNubError;
import com.pubnub.api.PubNubException;
import com.pubnub.api.PubNubUtil;
import com.pubnub.api.builder.PubNubErrorBuilder;
import com.pubnub.api.crypto.CryptoModule;
import com.pubnub.api.crypto.CryptoModuleKt;
import com.pubnub.api.managers.DuplicationManager;
import com.pubnub.api.managers.MapperManager;
import com.pubnub.api.models.consumer.files.PNDownloadableFile;
import com.pubnub.api.models.consumer.message_actions.PNMessageAction;
import com.pubnub.api.models.consumer.objects_api.channel.PNChannelMetadata;
import com.pubnub.api.models.consumer.objects_api.channel.PNChannelMetadataResult;
import com.pubnub.api.models.consumer.objects_api.membership.PNMembership;
import com.pubnub.api.models.consumer.objects_api.membership.PNMembershipResult;
import com.pubnub.api.models.consumer.objects_api.uuid.PNUUIDMetadata;
import com.pubnub.api.models.consumer.objects_api.uuid.PNUUIDMetadataResult;
import com.pubnub.api.models.consumer.pubsub.BasePubSubResult;
import com.pubnub.api.models.consumer.pubsub.PNEvent;
import com.pubnub.api.models.consumer.pubsub.PNMessageResult;
import com.pubnub.api.models.consumer.pubsub.PNPresenceEventResult;
import com.pubnub.api.models.consumer.pubsub.PNSignalResult;
import com.pubnub.api.models.consumer.pubsub.files.PNFileEventResult;
import com.pubnub.api.models.consumer.pubsub.message_actions.PNMessageActionResult;
import com.pubnub.api.models.consumer.pubsub.objects.ObjectPayload;
import com.pubnub.api.models.server.PresenceEnvelope;
import com.pubnub.api.models.server.PublishMetaData;
import com.pubnub.api.models.server.SubscribeMessage;
import com.pubnub.api.models.server.files.FileUploadNotification;
import com.pubnub.api.services.FilesService;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

@Slf4j
@AllArgsConstructor
public class SubscribeMessageProcessor {
    public static final int TYPE_MESSAGE = 0;
    private final int typeSignal = 1;
    private final int typeObject = 2;
    private final int typeMessageAction = 3;
    public static final int TYPE_FILES = 4;

    private final PubNub pubnub;
    private final DuplicationManager duplicationManager;

    private static final String PN_OTHER = "pn_other";

    @SuppressWarnings("deprecation")
    PNEvent processIncomingPayload(SubscribeMessage message) throws PubNubException {
        MapperManager mapper = this.pubnub.getMapper();

        String channel = message.getChannel();
        String subscriptionMatch = message.getSubscriptionMatch();
        PublishMetaData publishMetaData = message.getPublishMetaData();

        if (channel != null && channel.equals(subscriptionMatch)) {
            subscriptionMatch = null;
        }

        if (this.pubnub.getConfiguration().isDedupOnSubscribe()) {
            if (this.duplicationManager.isDuplicate(message)) {
                return null;
            }
            this.duplicationManager.addEntry(message);
        }

        if (message.getChannel().endsWith("-pnpres")) {
            PresenceEnvelope presencePayload = mapper.convertValue(message.getPayload(), PresenceEnvelope.class);

            String strippedPresenceChannel = null;
            String strippedPresenceSubscription = null;

            if (channel != null) {
                strippedPresenceChannel = PubNubUtil.replaceLast(channel, "-pnpres", "");
            }
            if (subscriptionMatch != null) {
                strippedPresenceSubscription = PubNubUtil.replaceLast(subscriptionMatch, "-pnpres", "");
            }

            JsonElement isHereNowRefresh = message.getPayload().getAsJsonObject().get("here_now_refresh");
            PNPresenceEventResult pnPresenceEventResult = PNPresenceEventResult.builder()
                    .event(presencePayload.getAction())
                    // deprecated
                    .actualChannel((subscriptionMatch != null) ? channel : null)
                    .subscribedChannel(subscriptionMatch != null ? subscriptionMatch : channel)
                    // deprecated
                    .channel(strippedPresenceChannel)
                    .subscription(strippedPresenceSubscription)
                    .state(presencePayload.getData())
                    .timetoken(publishMetaData.getPublishTimetoken())
                    .occupancy(presencePayload.getOccupancy())
                    .uuid(presencePayload.getUuid())
                    .timestamp(presencePayload.getTimestamp())
                    .join(getDelta(message.getPayload().getAsJsonObject().get("join")))
                    .leave(getDelta(message.getPayload().getAsJsonObject().get("leave")))
                    .timeout(getDelta(message.getPayload().getAsJsonObject().get("timeout")))
                    .hereNowRefresh(isHereNowRefresh != null && isHereNowRefresh.getAsBoolean())
                    .build();

            return pnPresenceEventResult;
        } else {
            JsonElement extractedMessage;
            PubNubError error = null;
            try {
                if (!message.supportsEncryption()) {
                    extractedMessage = message.getPayload();
                } else {
                    extractedMessage = tryDecryptMessage(message.getPayload(), pubnub.getCryptoModule(), pubnub.getMapper());
                }
            } catch (PubNubException e) {
                if (e.getPubnubError() == PubNubErrorBuilder.PNERROBJ_PNERR_CRYPTO_IS_CONFIGURED_BUT_MESSAGE_IS_NOT_ENCRYPTED) {
                    extractedMessage = message.getPayload();
                    error = e.getPubnubError();
                } else {
                    throw e;
                }
            }
            if (extractedMessage == null) {
                log.debug("unable to parse payload on #processIncomingMessages");
            }

            BasePubSubResult result = BasePubSubResult.builder()
                    // deprecated
                    .actualChannel((subscriptionMatch != null) ? channel : null)
                    .subscribedChannel(subscriptionMatch != null ? subscriptionMatch : channel)
                    // deprecated
                    .channel(channel)
                    .subscription(subscriptionMatch)
                    .timetoken(publishMetaData.getPublishTimetoken())
                    .publisher(message.getIssuingClientId())
                    .userMetadata(message.getUserMetadata())
                    .build();

            if (message.getType() == null) {
                return new PNMessageResult(result, extractedMessage, error);
            } else if (message.getType() == TYPE_MESSAGE) {
                return new PNMessageResult(result, extractedMessage, error);
            } else if (message.getType() == typeSignal) {
                return new PNSignalResult(result, extractedMessage);
            } else if (message.getType() == typeObject) {
                ObjectPayload objectPayload = mapper.convertValue(extractedMessage, ObjectPayload.class);
                String type = objectPayload.getType();
                if (canHandleObjectCallback(objectPayload)) {
                    switch (type) {
                        case "channel":
                            final PNChannelMetadataResult channelMetadataResult = new PNChannelMetadataResult(result,
                                    objectPayload.getEvent(), mapper.convertValue(objectPayload.getData(),
                                    PNChannelMetadata.class));
                            return channelMetadataResult;
                        case "membership":
                            final PNMembershipResult membershipResult = new PNMembershipResult(result,
                                    objectPayload.getEvent(), mapper.convertValue(objectPayload.getData(),
                                    PNMembership.class));
                            return membershipResult;
                        case "uuid":
                            final PNUUIDMetadataResult uuidMetadataResult = new PNUUIDMetadataResult(result,
                                    objectPayload.getEvent(),
                                    mapper.convertValue(objectPayload.getData(), PNUUIDMetadata.class));
                            return uuidMetadataResult;
                        default:
                    }
                }
            } else if (message.getType() == typeMessageAction) {
                ObjectPayload objectPayload = mapper.convertValue(extractedMessage, ObjectPayload.class);
                JsonObject data = objectPayload.getData().getAsJsonObject();
                if (!data.has("uuid")) {
                    data.addProperty("uuid", result.getPublisher());
                }
                return PNMessageActionResult.actionBuilder()
                        .result(result)
                        .event(objectPayload.getEvent())
                        .data(mapper.convertValue(data, PNMessageAction.class))
                        .build();
            } else if (message.getType() == TYPE_FILES) {
                FileUploadNotification event = mapper.convertValue(extractedMessage, FileUploadNotification.class);
                final JsonElement jsonMessage;
                if (event.getMessage() != null) {
                    jsonMessage = mapper.toJsonTree(event.getMessage());
                } else {
                    jsonMessage = JsonNull.INSTANCE;
                }

                return PNFileEventResult.builder()
                        .file(new PNDownloadableFile(event.getFile().getId(),
                                event.getFile().getName(),
                                buildFileUrl(message.getChannel(),
                                        event.getFile().getId(),
                                        event.getFile().getName())))
                        .message(event.getMessage())
                        .channel(message.getChannel())
                        .publisher(message.getIssuingClientId())
                        .timetoken(publishMetaData.getPublishTimetoken())
                        .jsonMessage(jsonMessage)
                        .build();
            }
        }
        return null;
    }

    public static JsonElement tryDecryptMessage(JsonElement input, CryptoModule cryptoModule, MapperManager mapper) throws PubNubException {
        // if we do not have a crypto module, there is no way to process the node; let's return.
        if (cryptoModule == null) {
            return input;
        }

        String inputText;
        String outputText;
        JsonElement outputObject;

        if (mapper.isJsonObject(input)) {
            if (mapper.hasField(input, PN_OTHER)) {
                inputText = mapper.elementToString(input, PN_OTHER);
            } else {
                throw logAndGetDecryptionException();
            }
        } else if (input.isJsonPrimitive() && input.getAsJsonPrimitive().isString()) {
            // String may represent not encrypted string or encrypted data. We will check this when decrypting.
            inputText = mapper.elementToString(input);
        } else {
            // Input represents some other Json structure, such as JsonArray
            throw logAndGetDecryptionException();
        }

        try {
            outputText = CryptoModuleKt.decryptString(cryptoModule, inputText);
            outputObject = mapper.fromJson(outputText, JsonElement.class);
        } catch (Exception e) {
            throw logAndGetDecryptionException();
        }

        // inject the decoded response into the payload
        if (mapper.isJsonObject(input) && mapper.hasField(input, PN_OTHER)) {
            JsonObject objectNode = mapper.getAsObject(input);
            mapper.putOnObject(objectNode, PN_OTHER, outputObject);
            outputObject = objectNode;
        }

        return outputObject;
    }

    private static PubNubException logAndGetDecryptionException() {
        PubNubError error = PubNubErrorBuilder.PNERROBJ_PNERR_CRYPTO_IS_CONFIGURED_BUT_MESSAGE_IS_NOT_ENCRYPTED;
        log.warn(error.getMessage());
        return new PubNubException(error.getMessage(), error, null, null, 0, null, null);
    }

    @SuppressWarnings("RegExpRedundantEscape")
    private final String formatFriendlyGetFileUrl = "%s" + FilesService.GET_FILE_URL.replaceAll("\\{.*?\\}", "%s");

    private String buildFileUrl(String channel, String fileId, String fileName) {
        String basePath = String.format(formatFriendlyGetFileUrl,
                pubnub.getBaseUrl(),
                pubnub.getConfiguration().getSubscribeKey(),
                channel,
                fileId,
                fileName);

        ArrayList<String> queryParams = new ArrayList<>();
        String authKey = pubnub.getConfiguration().getAuthKey();

        if (PubNubUtil.shouldSignRequest(pubnub.getConfiguration())) {
            int timestamp = pubnub.getTimestamp();
            String signature = generateSignature(pubnub.getConfiguration(), basePath, authKey, timestamp);
            queryParams.add(PubNubUtil.TIMESTAMP_QUERY_PARAM_NAME + "=" + timestamp);
            queryParams.add(PubNubUtil.SIGNATURE_QUERY_PARAM_NAME + "=" + signature);
        }

        if (authKey != null) {
            queryParams.add(PubNubUtil.AUTH_QUERY_PARAM_NAME + "=" + authKey);
        }

        if (queryParams.isEmpty()) {
            return basePath;
        } else {
            return basePath + "?" + PubNubUtil.joinString(queryParams, "&");
        }
    }

    private String generateSignature(PNConfiguration configuration, String url, String authKey, int timestamp) {
        HashMap<String, String> queryParams = new HashMap<>();
        if (authKey != null) {
            queryParams.put("auth", authKey);
        }
        return PubNubUtil.generateSignature(configuration,
                url,
                queryParams,
                "get",
                null,
                timestamp
        );
    }

    private boolean canHandleObjectCallback(final ObjectPayload objectPayload) {
        return objectPayload.getVersion().equals("2.0");
    }

    private List<String> getDelta(JsonElement delta) {
        List<String> list = new ArrayList<>();
        if (delta != null) {
            JsonArray jsonArray = delta.getAsJsonArray();
            for (int i = 0; i < jsonArray.size(); i++) {
                list.add(jsonArray.get(i).getAsString());
            }
        }

        return list;
    }
}

File: src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt
package com.pubnub.api.workers

import com.google.gson.JsonElement
import com.google.gson.JsonNull
import com.pubnub.api.PNConfiguration
import com.pubnub.api.PNConfiguration.Companion.isValid
import com.pubnub.api.PubNub
import com.pubnub.api.PubNubUtil
import com.pubnub.api.managers.DuplicationManager
import com.pubnub.api.models.consumer.MessageType
import com.pubnub.api.models.consumer.files.PNDownloadableFile
import com.pubnub.api.models.consumer.message_actions.PNMessageAction
import com.pubnub.api.models.consumer.pubsub.BasePubSubResult
import com.pubnub.api.models.consumer.pubsub.PNEvent
import com.pubnub.api.models.consumer.pubsub.PNMessageResult
import com.pubnub.api.models.consumer.pubsub.PNPresenceEventResult
import com.pubnub.api.models.consumer.pubsub.PNSignalResult
import com.pubnub.api.models.consumer.pubsub.files.PNFileEventResult
import com.pubnub.api.models.consumer.pubsub.message_actions.PNMessageActionResult
import com.pubnub.api.models.consumer.pubsub.objects.ObjectPayload
import com.pubnub.api.models.consumer.pubsub.objects.PNObjectEventMessage
import com.pubnub.api.models.consumer.pubsub.objects.PNObjectEventResult
import com.pubnub.api.models.server.PresenceEnvelope
import com.pubnub.api.models.server.SubscribeMessage
import com.pubnub.api.models.server.files.FileUploadNotification
import com.pubnub.api.services.FilesService
import com.pubnub.api.vendor.Crypto
import org.slf4j.LoggerFactory

internal class SubscribeMessageProcessor(
    private val pubnub: PubNub,
    private val duplicationManager: DuplicationManager
) {

    companion object {
        private const val PRESENCE_SUFFIX = "-pnpres"
        private const val PN_OTHER_FIELD = "pn_other"
    }

    private val log = LoggerFactory.getLogger("SubscribeMessageProcessor")

    fun processIncomingPayload(message: SubscribeMessage): PNEvent? {
        if (message.channel == null) {
            return null
        }

        val channel = message.channel
        var subscriptionMatch = message.subscriptionMatch
        val publishMetaData = message.publishMetaData

        if (channel == subscriptionMatch) {
            subscriptionMatch = null
        }

        if (pubnub.configuration.dedupOnSubscribe) {
            if (duplicationManager.isDuplicate(message)) {
                return null
            } else {
                duplicationManager.addEntry(message)
            }
        }

        if (message.channel.endsWith(PRESENCE_SUFFIX)) {
            val presencePayload = pubnub.mapper.convertValue(message.payload, PresenceEnvelope::class.java)
            val strippedPresenceChannel = PubNubUtil.replaceLast(channel, PRESENCE_SUFFIX, "")
            val strippedPresenceSubscription = subscriptionMatch?.let {
                PubNubUtil.replaceLast(it, PRESENCE_SUFFIX, "")
            }

            val isHereNowRefresh = message.payload?.asJsonObject?.get("here_now_refresh")

            return PNPresenceEventResult(
                event = presencePayload.action,
                uuid = presencePayload.uuid,
                timestamp = presencePayload.timestamp,
                occupancy = presencePayload.occupancy,
                state = presencePayload.data,
                channel = strippedPresenceChannel,
                subscription = strippedPresenceSubscription,
                timetoken = publishMetaData?.publishTimetoken,
                join = getDelta(message.payload?.asJsonObject?.get("join")),
                leave = getDelta(message.payload?.asJsonObject?.get("leave")),
                timeout = getDelta(message.payload?.asJsonObject?.get("timeout")),
                hereNowRefresh = isHereNowRefresh != null && isHereNowRefresh.asBoolean
            )
        } else {
            val extractedMessage = processMessage(message)

            if (extractedMessage == null) {
                log.debug("unable to parse payload on #processIncomingMessages")
            }

            val result = BasePubSubResult(
                channel = channel,
                subscription = subscriptionMatch,
                timetoken = publishMetaData?.publishTimetoken,
                userMetadata = message.userMetadata,
                publisher = message.issuingClientId
            )

            return when (message.pnMessageType) {
                is MessageType.Message -> {
                    PNMessageResult(
                        basePubSubResult = result,
                        message = extractedMessage!!,
                        spaceId = message.spaceId,
                        messageType = message.userMessageType ?: message.pnMessageType
                    )
                }

                is MessageType.Signal -> {
                    PNSignalResult(
                        basePubSubResult = result,
                        message = extractedMessage!!,
                        spaceId = message.spaceId,
                        messageType = message.userMessageType ?: message.pnMessageType
                    )
                }

                is MessageType.Object -> {
                    PNObjectEventResult(
                        result,
                        pubnub.mapper.convertValue(
                            extractedMessage, PNObjectEventMessage::class.java
                        )
                    )
                }

                is MessageType.MessageAction -> {
                    val objectPayload = pubnub.mapper.convertValue(extractedMessage, ObjectPayload::class.java)
                    val data = objectPayload.data.asJsonObject
                    if (!data.has("uuid")) {
                        data.addProperty("uuid", result.publisher)
                    }
                    PNMessageActionResult(
                        result = result,
                        event = objectPayload.event,
                        data = pubnub.mapper.convertValue(data, PNMessageAction::class.java)
                    )
                }

                is MessageType.File -> {
                    val fileUploadNotification = pubnub.mapper.convertValue(
                        extractedMessage, FileUploadNotification::class.java
                    )
                    PNFileEventResult(
                        channel = message.channel,
                        message = fileUploadNotification.message,
                        file = PNDownloadableFile(
                            id = fileUploadNotification.file.id,
                            name = fileUploadNotification.file.name,
                            url = buildFileUrl(
                                message.channel, fileUploadNotification.file.id, fileUploadNotification.file.name
                            )
                        ),
                        publisher = message.issuingClientId,
                        timetoken = result.timetoken,
                        jsonMessage = fileUploadNotification.message?.let { pubnub.mapper.toJsonTree(it) }
                            ?: JsonNull.INSTANCE,
                        spaceId = message.spaceId,
                        messageType = message.userMessageType ?: message.pnMessageType,
                        userMetadata = message.userMetadata
                    )
                }
            }
        }
    }

    private val formatFriendlyGetFileUrl = "%s" + FilesService.GET_FILE_URL.replace("\\{.*?\\}".toRegex(), "%s")

    private fun buildFileUrl(channel: String, fileId: String, fileName: String): String {
        val basePath: String = java.lang.String.format(
            formatFriendlyGetFileUrl, pubnub.baseUrl(), pubnub.configuration.subscribeKey, channel, fileId, fileName
        )
        val queryParams = ArrayList<String>()
        val authKey = if (pubnub.configuration.authKey.isValid()) pubnub.configuration.authKey else null

        if (PubNubUtil.shouldSignRequest(pubnub.configuration)) {
            val timestamp: Int = pubnub.timestamp()
            val signature: String = generateSignature(pubnub.configuration, basePath, authKey, timestamp)
            queryParams.add(PubNubUtil.TIMESTAMP_QUERY_PARAM_NAME + "=" + timestamp)
            queryParams.add(PubNubUtil.SIGNATURE_QUERY_PARAM_NAME + "=" + signature)
        }

        authKey?.run { queryParams.add(PubNubUtil.AUTH_QUERY_PARAM_NAME + "=" + authKey) }

        return if (queryParams.isEmpty()) {
            basePath
        } else {
            "$basePath?${queryParams.joinToString(separator = "&")}"
        }
    }

    private fun generateSignature(
        configuration: PNConfiguration,
        url: String,
        authKey: String?,
        timestamp: Int
    ): String {
        val queryParams = mutableMapOf<String, String>()
        if (authKey != null) {
            queryParams["auth"] = authKey
        }
        return PubNubUtil.generateSignature(
            configuration, url, queryParams, "get", null, timestamp
        )
    }

    private fun processMessage(subscribeMessage: SubscribeMessage): JsonElement? {
        val input = subscribeMessage.payload

        // if we do not have a crypto key, there is no way to process the node; let's return.
        if (!pubnub.configuration.cipherKey.isValid()) {
            return input
        }

        // if the message couldn't possibly be encrypted in the first place, there is no way to process the node;
        // let's return.
        if (!subscribeMessage.supportsEncryption()) {
            return input
        }

        val crypto = Crypto(
            pubnub.configuration.cipherKey, pubnub.configuration.useRandomInitializationVector
        )

        val inputText = if (pubnub.mapper.isJsonObject(input!!) && pubnub.mapper.hasField(input, PN_OTHER_FIELD)) {
            pubnub.mapper.elementToString(input, PN_OTHER_FIELD)
        } else {
            pubnub.mapper.elementToString(input)
        }

        val outputText = crypto.decrypt(inputText!!)
        var outputObject = pubnub.mapper.fromJson(outputText, JsonElement::class.java)

        if (pubnub.mapper.isJsonObject(input) && pubnub.mapper.hasField(input, PN_OTHER_FIELD)) {
            val objectNode = pubnub.mapper.getAsObject(input)
            pubnub.mapper.putOnObject(objectNode, PN_OTHER_FIELD, outputObject)
            outputObject = objectNode
        }

        return outputObject
    }

    private fun getDelta(delta: JsonElement?): List<String> {
        val list = mutableListOf<String>()
        delta?.let {
            it.asJsonArray.forEach { item: JsonElement? ->
                item?.let {
                    list.add(it.asString)
                }
            }
        }
        return list
    }
}
